Feat: Support specifying `publishVersion` when creating or updating oam applications

iambocai opened the pull request:
### Description of your changes

<!--

Briefly describe what this pull request does. We love pull requests that resolve an open KubeVela issue. If yours does, you
can uncomment the below line to indicate which issue your PR fixes, for example
"Fixes #500":

-->
1. Support specifying `publishVersion` when creating or updating oam applications **via API**
2. update version in `metadata.yaml` from `v1.8.0-rc.3` to `v1.8.0-rc.4` 


I have:

- [x] Read and followed KubeVela's [contribution process](https://github.com/oam-dev/kubevela/blob/master/contribute/create-pull-request.md).
- [x] [Related Docs](https://github.com/oam-dev/kubevela.io) updated properly. In a new feature or configuration option, an update to the documentation is necessary.
- [x] Run `yarn lint` to ensure the frontend changes are ready for review.
- [x] Run `make reviewable`to ensure the server changes are ready for review.
- [x] Added `backport release-x.y` labels to auto-backport this PR if necessary.

### Special notes for your reviewer

<!--

Be sure to direct your reviewers'
attention to anything that needs special consideration.
-->

1. Because the requirement to specify publishVersion usually only exists during API calls, supporting users to specify this parameter on the front-end can actually lead to unnecessary understanding costs, so we **didn't** add support for specifying relevant parameters on the UX front-end.
2. The VelaUX version indicated in the `metadata.yaml` file is `1.8.0-rc3`, but in reality, the available version has already reached `1.9.3`. Should I also change this to `1.9.3` or `1.9.4` (instead of `v1.8.0-rc.4`)?

Commit:
feat(oam_application): Support specifying publishVersion when creating or updating oam applications

Signed-off-by: 上郡 <shangjun.csb@alibaba-inc.com>

Commit:
feat(oam_application): Support specifying publishVersion when creating or updating oam applications

Signed-off-by: 上郡 <shangjun.csb@alibaba-inc.com>

Commit:
feat(apidoc): Add swagger-3.0.json

Signed-off-by: 上郡 <shangjun.csb@alibaba-inc.com>

Commit:
fix #713: duplicate field 'runtime_cluster' in swagger.json

Signed-off-by: 上郡 <shangjun.csb@alibaba-inc.com>

iambocai commented:
Additionally, as I urgently need this feature in my production environment, I would like to release a new version of VelaUX

Commit:
fix: add annotations on update

Signed-off-by: 上郡 <shangjun.csb@alibaba-inc.com>

Commit:
fix: dryrun annotations fix

Signed-off-by: 上郡 <shangjun.csb@alibaba-inc.com>

FogDong reviewed (commented):
Since this PR only contains API changes, does it mean that we can not specify publishVersion in the frontend?

iambocai commented:
> Since this PR only contains API changes, does it mean that we can not specify publishVersion in the frontend?

Yes, I am considering this: if users are supported to fill in PublishVersion on a web page, then for web users, they need to understand what PublishVersion is, why uniqueness needs to be guaranteed, and its relationship with Revision (currently, without specifying PublishVersion, the default value of PublishVersion is the same as Revision), This will bring two problems: firstly, it brings some learning costs to web (usually beginner) users; Secondly, they are easy to fill in duplicates (considering scenarios where multiple people are simultaneously deploying the same application on a web page), and the server needs to support pre-validation support.

Of course, if necessary, I can also add front-end support.

iambocai commented on docs/contributing/velaux.md:
````diff
@@ -103,6 +103,9 @@ make e2e-server-test
 make build-swagger
 ```
 
+if you prefer v3 version of OpenAPI schema, please convert it 
````
OK, no problem, just wait a moment~

iambocai replied:
> Please separate openAPI v3 update to another PR. Let's focus on publishVersion in this one.

Done

Commit:
feat(oam_application): Support specifying publishVersion when creating or updating oam applications

Signed-off-by: 上郡 <shangjun.csb@alibaba-inc.com>

Commit:
rebase commit, pick only publishversion related updates

* 'feature/app_publish_version' of github.com:iambocai/velaux:
  fix: dryrun annotations fix
  fix: add annotations on update
  fix #713: duplicate field 'runtime_cluster' in swagger.json
  feat(apidoc): Add swagger-3.0.json
  feat(oam_application): Support specifying publishVersion when creating or updating oam applications

Commit:
rebase 'feature/app_publish_version' of github.com:iambocai/velaux

* 'feature/app_publish_version' of github.com:iambocai/velaux:
  feat(oam_application): Support specifying publishVersion when creating or updating oam applications
# 请输入一个提交信息以解释此合并的必要性，尤其是将一个更新后的上游分支
# 合并到主题分支。
#
# 以 '#' 开始的行将被忽略，而空的提交说明将终止提交。

FogDong reviewed (approved):
Generally LGTM if this feature is only supported in API.
Any insights? cc @barnettZQG

barnettZQG commented:
@iambocai I think this change don't impact on the UI. LGTM after passed the CI checks.

iambocai commented:
> @iambocai I think this change don't impact on the UI. LGTM after passed the CI checks.


thanks for your reply, I saw some test faild in ci, but all of them seems to be environment problems, may be we should fix them in another commit？

VelaUX APIServer Test / server-e2e-tests (v1.21) (pull_request) —— No matching K3s versions were found.
![image](https://github.com/kubevela/velaux/assets/2259715/e0b523c5-4ada-42a0-be19-14f3f02bf221)
the avaliable k3s versions started at v22, but we request v21.

Commit:
update test k8s-version

Signed-off-by: 上郡 <shangjun.csb@alibaba-inc.com>

Commit:
update Kubebuilder k8s version

Signed-off-by: 上郡 <shangjun.csb@alibaba-inc.com>

iambocai commented:
> > @iambocai I think this change don't impact on the UI. LGTM after passed the CI checks.
> 
> thanks for your reply, I saw some test faild in ci, but all of them seems to be environment problems, may be we should fix them in another commit？
> 
> VelaUX APIServer Test / server-e2e-tests (v1.21) (pull_request) —— No matching K3s versions were found. ![image](https://private-user-images.githubusercontent.com/2259715/285733050-e0b523c5-4ada-42a0-be19-14f3f02bf221.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTEiLCJleHAiOjE3MDEwNzI0MzIsIm5iZiI6MTcwMTA3MjEzMiwicGF0aCI6Ii8yMjU5NzE1LzI4NTczMzA1MC1lMGI1MjNjNS00YWRhLTQyYTAtYmUxOS0xNGYzZjAyYmYyMjEucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQUlXTkpZQVg0Q1NWRUg1M0ElMkYyMDIzMTEyNyUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyMzExMjdUMDgwMjEyWiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9OGQ2OTZlOTMyNWVlYTlmYjc2MTFlZDhiYjlmOTkzOTU0ZjUyM2ZiMTZkZmQzNGE4OWEzOWRhMThjY2VjZTc5MyZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.eWUAnIovOuRvWj68rrELBqlkyeDO9Xagu-fOknfE0OY) the avaliable k3s versions started at v22, but we request v21.

I have upgrade the minor  server-e2e-tests k8s version to 1.22 and passed the check, but ci still expect v1.21 test's result, may be we should modify git settings too?
![image](https://github.com/kubevela/velaux/assets/2259715/1bc42bbf-03f4-4ad4-a85a-46a247dabe9f)

wonderflow commented on .github/workflows/server-test.yml:
```diff
@@ -91,7 +91,7 @@ jobs:
           version: 3.1.0
           kubebuilderOnly: false
-          kubernetesVersion: v1.21.2
+          kubernetesVersion: v1.22.17
```
we'd better revert the change back to 1.21

iambocai replied:
Then we should find some way to make the k3s setup works again, currently [the k3s setup script](https://github.com/nolar/setup-k3d-k3s/blob/v1/action.sh) only support latest 200 versions ( which `>= v1.24.7-rc3+k3s1`, and continue growth day by day):

![image](https://github.com/kubevela/velaux/assets/2259715/ffcc47aa-445d-4ed9-9340-0d7670063a2a)

```shell
curl --silent --fail --location "${authz[@]-}"  "https://api.github.com/repos/k3s-io/k3s/releases?per_page=999&page=2"
````
![image](https://github.com/kubevela/velaux/assets/2259715/e95957f2-d2ac-4792-aefc-23e95656747a)

maybe we should find another k3s setup script to workaround this issue.
